File: dart/lib/src/twitter/create_verified_twitter_registry.dart
import 'dart:typed_data';

import '../constants/addresses.dart';
import '../instructions/instruction_types.dart';
import '../int.dart';
import '../rpc/rpc_client.dart';
import '../utils/get_domain_key_sync.dart';
import 'create_reverse_twitter_registry.dart';

/// Create verified Twitter registry instruction
///
/// This function mirrors js/src/twitter/createVerifiedTwitterRegistry.ts
///
/// Signed by the authority, the payer and the verified pubkey
///
/// [connection] - RPC connection
/// [twitterHandle] - The Twitter handle
/// [verifiedPubkey] - The verified public key
/// [space] - The space that the user will have to write data into the verified registry
/// [payerKey] - The payer's public key
///
/// Returns list of transaction instructions
Future<List<TransactionInstruction>> createVerifiedTwitterRegistry(
  RpcClient connection,
  String twitterHandle,
  String verifiedPubkey,
  int space,
  String payerKey,
) async {
  // Create user facing registry
  final hashedTwitterHandle = getHashedNameSync(twitterHandle);
  final twitterHandleRegistryKey = getNameAccountKeySync(
    hashedTwitterHandle,
    nameParent: twitterRootParentRegistryAddress,
  );

  // Calculate rent exemption for the account
  // This mirrors the JavaScript SDK's getMinimumBalanceForRentExemption call
  // Using approximate calculation similar to other instructions in the codebase
  final accountSize =
      space + 96; // space + NameRegistryState.HEADER_LEN equivalent
  final rentExemptionLamports =
      accountSize * 6960; // Approximate rent exemption calculation

  final instructions = <TransactionInstruction>[];

  // Create the main user-facing registry instruction matching JavaScript SDK's createInstruction
  final createInstruction = TransactionInstruction(
    programAddress: nameProgramAddress,
    accounts: [
      const AccountMeta(
        address: systemProgramAddress,
        role: AccountRole.readonly,
      ),
      AccountMeta(
        address: payerKey,
        role: AccountRole.writableSigner,
      ),
      AccountMeta(
        address: await twitterHandleRegistryKey,
        role: AccountRole.writable,
      ),
      AccountMeta(
        address: verifiedPubkey,
        role: AccountRole.readonly,
      ),
      // Optional class key - unused for user-facing registries
      const AccountMeta(
        address:
            '11111111111111111111111111111111', // PublicKey(Buffer.alloc(32)) equivalent
        role: AccountRole.readonly,
      ),
      // Parent key
      const AccountMeta(
        address: twitterRootParentRegistryAddress,
        role: AccountRole.readonly,
      ),
      // Parent owner (Twitter verification authority acts as owner of the parent)
      const AccountMeta(
        address: twitterVerificationAuthority,
        role: AccountRole.readonlySigner,
      ),
    ],
    data: _buildCreateInstructionData(
        hashedTwitterHandle, rentExemptionLamports, space),
  );
  instructions.add(createInstruction);

  // Create the reverse Twitter registry
  final reverseInstructions = await createReverseTwitterRegistry(
    connection,
    twitterHandle,
    await twitterHandleRegistryKey,
    verifiedPubkey,
    payerKey,
  );
  instructions.addAll(reverseInstructions);

  return instructions;
}

/// Build create instruction data matching JavaScript SDK implementation
///
/// This mirrors js/src/instructions/createInstruction.ts format:
/// [0] - Instruction tag (0 for create)
/// [1-4] - Hashed name length (u32, little-endian)
/// [5-36] - Hashed name (32 bytes)
/// [37-44] - Lamports (u64, little-endian)
/// [45-48] - Space (u32, little-endian)
Uint8List _buildCreateInstructionData(
  Uint8List hashedName,
  int lamports,
  int space,
) {
  final data = <int>[
    // Instruction tag (0 for create)
    0,
  ];

  // Add hashed name length (4 bytes, little-endian)
  final nameLength = Numberu32(hashedName.length);
  data.addAll(nameLength.toBuffer());

  // Add hashed name (32 bytes)
  data.addAll(hashedName);

  // Add lamports (8 bytes, little-endian)
  final lamportsBuffer = Numberu64(lamports);
  data.addAll(lamportsBuffer.toBuffer());

  // Add space (4 bytes, little-endian)
  final spaceBuffer = Numberu32(space);
  data.addAll(spaceBuffer.toBuffer());

  return Uint8List.fromList(data);
}

File: js/src/twitter/createVerifiedTwitterRegistry.ts
import {
  Connection,
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import {
  NAME_PROGRAM_ID,
  TWITTER_VERIFICATION_AUTHORITY,
  TWITTER_ROOT_PARENT_REGISTRY_KEY,
} from "../constants";
import { createInstruction } from "../instructions/createInstruction";
import { NameRegistryState } from "../state";
import { getHashedNameSync } from "../utils/getHashedNameSync";
import { getNameAccountKeySync } from "../utils/getNameAccountKeySync";
import { Numberu32, Numberu64 } from "../int";

import { createReverseTwitterRegistry } from "./createReverseTwitterRegistry";

/**
 * Builds instructions to create a verified Twitter handle registry and its reverse registry.
 *
 * The authority, payer, and verified public key must sign the resulting instructions.
 *
 * @param connection Solana RPC connection used to calculate rent-exemption costs.
 * @param twitterHandle Twitter handle to verify and register.
 * @param verifiedPubkey Public key associated with the verified handle.
 * @param space Number of bytes available in the user-facing registry.
 * @param payerKey Signer that funds both registry accounts.
 * @returns Instructions that create the handle and reverse registries.
 *
 * @example
 * ```ts
 * const instructions = await createVerifiedTwitterRegistry(connection, "bonfida", owner, 128, payer);
 * ```
 */
export async function createVerifiedTwitterRegistry(
  connection: Connection,
  twitterHandle: string,
  verifiedPubkey: PublicKey,
  space: number, // The space that the user will have to write data into the verified registry
  payerKey: PublicKey,
): Promise<TransactionInstruction[]> {
  // Create user facing registry
  const hashedTwitterHandle = getHashedNameSync(twitterHandle);
  const twitterHandleRegistryKey = getNameAccountKeySync(
    hashedTwitterHandle,
    undefined,
    TWITTER_ROOT_PARENT_REGISTRY_KEY,
  );

  const lamports = await connection.getMinimumBalanceForRentExemption(
    space + NameRegistryState.HEADER_LEN,
  );

  let instructions = [
    createInstruction(
      NAME_PROGRAM_ID,
      SystemProgram.programId,
      twitterHandleRegistryKey,
      verifiedPubkey,
      payerKey,
      hashedTwitterHandle,
      new Numberu64(lamports),
      new Numberu32(space),
      undefined,
      TWITTER_ROOT_PARENT_REGISTRY_KEY,
      TWITTER_VERIFICATION_AUTHORITY, // Twitter authority acts as owner of the parent for all user-facing registries
    ),
  ];

  instructions = instructions.concat(
    await createReverseTwitterRegistry(
      connection,
      twitterHandle,
      twitterHandleRegistryKey,
      verifiedPubkey,
      payerKey,
    ),
  );

  return instructions;
}
